common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java

```diff
@@ -430,7 +430,7 @@ public static int execBinary(final UTF8String string, final UTF8String substring
     }
     public static int execLowercase(final UTF8String string, final UTF8String substring,
         final int start) {
-      return string.toLowerCase().indexOf(substring.toLowerCase(), start);
```

Comment:
to confirm, the previous implementation here is correct, right?

Reply:
no, unfortunately it's not - while it works fine for ASCII, it actually gives wrong results in some special cases featuring conditional case mapping, when a character has a lowercase equivalent that consists of multiple characters, or is found at a particular place in the string (context-awareness)

Reply:
so as part of this PR, we actually changed the core definition of string-searching in UTF8_BINARY_LCASE, i.e. what it means for one substring (_pattern_) to be found in another string (_target_) under UTF8_BINARY_LCASE

in the old implementation, `contains("İ", "i")` would return `true` - however, this behaviour is incorrect because it relies on the fact that `substr(lower("İ"))` == `"i"` (incorrect, old implementation), instead of `lower(substr("İ"))` == `"i"` (correct, new implementation)